rs-matter/src/transport/core.rs

```diff
@@ -15,79 +15,198 @@
-
-    const TX_INIT: [TxBuf; MAX_EXCHANGES] = [Self::TX_ELEM; MAX_EXCHANGES];
-    const RX_INIT: [RxBuf; MAX_EXCHANGES] = [Self::RX_ELEM; MAX_EXCHANGES];
-    const SX_INIT: [SxBuf; MAX_EXCHANGES + 1] = [Self::SX_ELEM; MAX_EXCHANGES + 1];
-
+impl<'m> TransportMgr<'m> {
     #[inline(always)]
-    pub const fn new() -> Self {
+    pub(crate) const fn new(mdns: MdnsImpl<'m>, epoch: Epoch, rand: Rand) -> Self {
         Self {
-            tx: Self::TX_INIT,
-            rx: Self::RX_INIT,
-            sx: Self::SX_INIT,
+            rx: IfMutex::new(Packet::new()),
+            tx: IfMutex::new(Packet::new()),
+            dropped: Notification::new(),
+            session_mgr: RefCell::new(SessionMgr::new(epoch, rand)),
+            mdns,
         }
     }
-}
 
-impl<'a> Matter<'a> {
+    #[cfg(all(feature = "large-buffers", feature = "alloc"))]
+    pub fn initialize_buffers(&self) -> Result<(), Error> {
+        let mut rx = self.rx.try_lock().map_err(|_| ErrorCode::InvalidState)?;
+        let mut tx = self.tx.try_lock().map_err(|_| ErrorCode::InvalidState)?;
+
+        if rx.buf.0.is_none() {
+            rx.buf.0 = Some(alloc::boxed::Box::new(heapless::Vec::new()));
+        }
+
+        if tx.buf.0.is_none() {
+            tx.buf.0 = Some(alloc::boxed::Box::new(heapless::Vec::new()));
+        }
+
+        Ok(())
+    }
+
+    #[cfg(not(all(feature = "large-buffers", feature = "alloc")))]
+    pub fn initialize_buffers(&self) -> Result<(), Error> {
+        // No-op, as buffers are allocated inline
+        Ok(())
+    }
+
+    pub fn reset(&self) {
+        self.session_mgr.borrow_mut().reset();
+        self.mdns.reset();
+    }
+
+    pub(crate) async fn initiate<'a>(
+        &'a self,
+        matter: &'a Matter<'a>,
+        node_id: u64,
+        secure: bool,
+    ) -> Result<Exchange<'_>, Error> {
+        let mut session_mgr = self.session_mgr.borrow_mut();
+
+        session_mgr
+            .get_for_node(node_id, secure)
+            .ok_or(ErrorCode::NoSession)?;
+
+        let exch_id = session_mgr.get_next_exch_id();
+
+        let session = session_mgr.get_for_node(node_id, secure).unwrap();
+        let exch_index = session
+            .add_exch(exch_id, Role::Initiator(Default::default()))
+            .ok_or(ErrorCode::NoSpaceExchanges)?;
+
+        let id = ExchangeId::new(session.id, exch_index);
+
+        info!("Exchange {id}: Initiated");
+
+        Ok(Exchange::new(id, matter))
+    }
+
+    pub(crate) async fn accept_if<'a, F>(
+        &'a self,
+        matter: &'a Matter<'a>,
+        mut f: F,
+    ) -> Result<Exchange<'_>, Error>
+    where
+        F: FnMut(&Session, &ExchangeState, &Packet<MAX_RX_BUF_SIZE>) -> bool,
+    {
+        let exchange = self
+            .with_locked(&self.rx, |packet| {
+                let mut session_mgr = self.session_mgr.borrow_mut();
+
+                if let Some(session) = session_mgr.get_for_rx(&packet.peer, &packet.header.plain) {
```

Comment:
consider early-return on these to unnest as this goes 3-levels of if deep and it is already in a closure in a call.